In [110]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

# **DATASET INTO DATAFRAME**

In [111]:
UK = pd.read_csv('Data-set(UK)\\uk_road_accident.csv')

In [112]:
UK

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

# Descrivtive Analysis

In [113]:
UK.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [114]:
UK.describe().T

count       mean       std        min        25%  \
Latitude              660654.0  52.553866  1.406922  49.914430  51.490690   
Longitude             660653.0  -1.431210  1.383330  -7.516225  -2.332291   
Number_of_Casualties  660679.0   1.357040  0.824847   1.000000   1.000000   
Number_of_Vehicles    660679.0   1.831255  0.715269   1.000000   1.000000   

                            50%        75%        max  
Latitude              52.315641  53.453452  60.757544  
Longitude             -1.411667  -0.232869   1.762010  
Number_of_Casualties   1.000000   1.000000  68.000000  
Number_of_Vehicles     2.000000   2.000000  32.000000

# NULL

In [115]:
UK.isnull()

Index  Accident_Severity  Accident Date  Latitude  Light_Conditions  \
0       False              False          False     False             False   
1       False              False          False     False             False   
2       False              False          False     False             False   
3       False              False          False     False             False   
4       False              False          False     False             False   
...       ...                ...            ...       ...               ...   
660674  False              False          False     False             False   
660675  False              False          False     False             False   
660676  False              False          False     False             False   
660677  False              False          False     False             False   
660678  False              False          False     False             False   

        District Area  Longitude  Number_of_Casualties  Number_of_Vehicles  \
0               False      False                 False               False   
1               False      False                 False               False   
2               False      False                 False               False   
3               False      False                 False               False   
4               False      False                 False               False   
...               ...        ...                   ...                 ...   
660674          False      False                 False               False   
660675          False      False                 False               False   
660676          False      False                 False               False   
660677          False      False                 False               False   
660678          False      False                 False               False   

        Road_Surface_Conditions  Road_Type  Urban_or_Rural_Area  \
0                         False      False                False   
1                         False      False                False   
2                         False       True                False   
3                         False      False                False   
4                         False       True                False   
...                         ...        ...                  ...   
660674                    False      False                False   
660675                    False      False                False   
660676                    False      False                False   
660677                    False      False                False   
660678                    False      False                False   

        Weather_Conditions  Vehicle_Type  
0                    False         False  
1                    False         False  
2                     True         False  
3                    False         False  
4                     True         False  
...                    ...           ...  
660674               False         False  
660675               False         False  
660676               False         False  
660677               False         False  
660678               False         False  

[660679 rows x 14 columns]

In [116]:
UK.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [117]:
UK['Latitude'].mean()

np.float64(52.553865761110956)

In [118]:
UK['Longitude'].mean()

np.float64(-1.431210368502073)

In [119]:
UK['Road_Surface_Conditions'].mode()

0    Dry
Name: Road_Surface_Conditions, dtype: object

In [120]:
UK['Road_Type'].mode()

0    Single carriageway
Name: Road_Type, dtype: object

In [121]:
UK['Urban_or_Rural_Area'].mode()

0    Urban
Name: Urban_or_Rural_Area, dtype: object

In [122]:
UK['Weather_Conditions'].mode()

0    Fine no high winds
Name: Weather_Conditions, dtype: object

In [123]:
UK['Latitude'] = UK['Latitude'].fillna(UK['Latitude'].mean())
UK['Longitude'] = UK['Longitude'].fillna(UK['Longitude'].mean())
UK['Road_Surface_Conditions'] = UK['Road_Surface_Conditions'].fillna(UK['Road_Surface_Conditions'].mode()[0])
UK['Road_Type'] = UK['Road_Type'].fillna(UK['Road_Type'].mode()[0])
UK['Urban_or_Rural_Area'] = UK['Urban_or_Rural_Area'].fillna(UK['Urban_or_Rural_Area'].mode()[0])
UK['Weather_Conditions'] = UK['Weather_Conditions'].fillna(UK['Weather_Conditions'].mode()[0])

In [124]:
UK.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

# CHANGE THE DATA TYPE OF A COLUMN

In [125]:
UK.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660679 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660679 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  660679 non-null  object 
 10  Road_Type                660679 non-null  object 
 11  Urban_or_Rural_Area      660679 non-null  object 
 12  Weather_Conditions       660679 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [128]:
UK['Index'] = UK['Index'].astype('category')
UK['Accident_Severity'] = UK['Accident_Severity'].astype('category')
UK['Accident Date'] = UK['Accident Date'].astype('category')
UK['Light_Conditions'] = UK['Light_Conditions'].astype('category')
UK['District Area'] = UK['District Area'].astype('category')
UK['Road_Surface_Conditions'] = UK['Road_Surface_Conditions'].astype('category')
UK['Road_Type'] = UK['Road_Type'].astype('category')
UK['Urban_or_Rural_Area'] = UK['Urban_or_Rural_Area'].astype('category')
UK['Weather_Conditions'] = UK['Weather_Conditions'].astype('category')
UK['Vehicle_Type'] = UK['Vehicle_Type'].astype('category')

In [129]:
UK.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Index                    660679 non-null  category
 1   Accident_Severity        660679 non-null  category
 2   Accident Date            660679 non-null  category
 3   Latitude                 660679 non-null  float64 
 4   Light_Conditions         660679 non-null  category
 5   District Area            660679 non-null  category
 6   Longitude                660679 non-null  float64 
 7   Number_of_Casualties     660679 non-null  int64   
 8   Number_of_Vehicles       660679 non-null  int64   
 9   Road_Surface_Conditions  660679 non-null  category
 10  Road_Type                660679 non-null  category
 11  Urban_or_Rural_Area      660679 non-null  category
 12  Weather_Conditions       660679 non-null  category
 13  Vehicle_Type             660679 non-null  ca

# QUESTIONS AND INSIGHTS

## 1. How many Fatal, Serious, and Slight accidents were reported in the accident severity? Which kind of accident has the highest number and which has the lowest?

In [ ]:
Accident_df = UK['Accident_Severity'].value_counts()

In [ ]:
Accident_df

### ***Insight:*** According to the report, the total number of Fatal accidents is **8,661**, which is the lowest. Serious accidents are **88,217**, while Slight accidents have the highest total of **563,801**.|

# 2. How many accidents happened on 26-08-2019?

In [131]:
UK[UK['Accident Date']=='28-02-2022']['Accident_Severity'].value_counts()

Accident_Severity
Slight     208
Serious     35
Fatal        9
Name: count, dtype: int64

## ***Insight:*** On 28-02-2022, most accidents were Slight, fewer were Serious and Fatal.

# 3. What is the average number of vehicles involved in an accident?

In [134]:
avg_vehicles = UK['Number_of_Vehicles'].mean()

In [135]:
avg_vehicles

np.float64(1.8312554205597575)

## ***Insight:*** The average number of vehicles involved in an accident is about **1.83**.

# 4. According to data, what is the usual weather condition during an accidents?

In [138]:
weather_cdn = UK['Weather_Conditions'].mode()[0]

In [139]:
weather_cdn

'Fine no high winds'

## ***Insights:*** Most accidents happen when the weather is **Fine with no high winds**.

# 5. In what area does serious accident usually happened?

In [143]:
serious_acc = UK[UK['Accident_Severity'] == 'Serious']

In [144]:
serious_acc

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
5       200701BS67159           Serious    18-09-2019  51.497750   
...               ...               ...           ...        ...   
660629  201091NL00686           Serious     11/2/2022  57.687933   
660645  201091NL05842           Serious    18-10-2022  60.162660   
660672  201091NM01085           Serious    29-01-2022  57.170725   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
5                          Daylight  Kensington and Chelsea  -0.192556   
...                             ...                     ...        ...   
660629                     Daylight        Shetland Islands  -4.176860   
660645                     Daylight        Shetland Islands  -1.165232   
660672                     Daylight                Highland  -3.852421   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
5                          2                   3                     Dry   
...                      ...                 ...                     ...   
660629                     1                   2                     Dry   
660645                     1                   1                     Dry   
660672                     1                   2                    Snow   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
5       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660629  Single carriageway               Rural      Fine + high winds   
660645  Single carriageway               Rural      Fine + high winds   
660672  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
5                                        Car  
...                                      ...  
660629                                   Car  
660645                                   Car  
660672                                   Car  
660677                 Motorcycle over 500cc  
660

In [145]:
serious_acc['District Area'].value_counts()

District Area
Birmingham                   1474
Leeds                        1147
Westminster                   926
Bradford                      802
Sheffield                     749
                             ... 
Oswestry                       23
Orkney Islands                 19
Clackmannanshire               18
Shetland Islands               15
London Airport (Heathrow)       9
Name: count, Length: 422, dtype: int64

## ***Insights*** Most Serious accidents happened in **Birmingham** with **1,474 accidents**, followed by Leeds and Westminster. The fewest Serious accidents happened in London Airport (Heathrow) and Shetland Islands.

# 6. What type of road do accidents usually happen on?

In [151]:
road_acc = UK['Road_Type']

In [152]:
road_acc

0         Single carriageway
1         Single carriageway
2         Single carriageway
3         Single carriageway
4         Single carriageway
                 ...        
660674    Single carriageway
660675    Single carriageway
660676    Single carriageway
660677    Single carriageway
660678      Dual carriageway
Name: Road_Type, Length: 660679, dtype: category
Categories (5, object): ['Dual carriageway', 'One way street', 'Roundabout', 'Single carriageway', 'Slip road']

In [153]:
UK['Road_Type'].value_counts()

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

## ***Insight:*** Accidents usually happen on **Single carriageways**, because most accidents in the dataset happened on this type of road.

# 7. How many casualties usually happen per accident?

In [154]:
UK['Number_of_Casualties'].sum() / UK['Number_of_Casualties'].count()

np.float64(1.357040257068864)

## ***Insight:*** On average, there are about 1.36 casualties in each accident.

# 8. How many accidents happened in each Urban or Rural area?

In [156]:
UK.groupby('Urban_or_Rural_Area')['Index'].count()

Urban_or_Rural_Area
Rural          238990
Unallocated        11
Urban          421678
Name: Index, dtype: int64

## ***Insight:*** Most accidents happened in ***Urban areas***, fewer in Rural areas.

# 9. How many accidents happened on Roundabouts?

In [157]:
UK[UK['Road_Type'] == 'Roundabout']['Index'].count()

np.int64(43992)

## ***Insight:*** There were ***43,992*** accidents on Roundabouts.

# 10. Which district has the most slight accident?

In [160]:
UK[UK['Accident_Severity']=='Slight']['District Area'].mode()[0]

'Birmingham'

## ***Insight:*** Most ***Slight*** accidents happened in Birmingham.

# 11. How many Serious accidents happened in Leeds?

In [161]:
UK[(UK['Accident_Severity']=='Serious') & (UK['District Area']=='Leeds')]['Index'].count()

np.int64(1147)

## ***Insights:*** Leeds had recorded ***1,147** Serious accidents.

# 12. What is the most common accident?

In [164]:
UK['Accident_Severity'].mode()[0]

'Slight'

## ***Insights:*** Most accidents are ***Slight***

# 13. How many accidents happened in Bradford?

In [165]:
UK[UK['District Area']=='Bradford']['Index'].count()

np.int64(6212)

## ***Insights:*** There are total of ***802*** accidents in Bradford.

# 14. How many accidents happened in Westminster?

In [166]:
UK[UK['District Area']=='Westminster']['Index'].count()

np.int64(5706)

## ***Insights:*** There are ***5,706*** accidents in Westminster.

# 15. What is the usual type of vehicle involed in accidents?

In [170]:
UK['Vehicle_Type'].mode()[0]

'Car'

## ***Insights:*** Most accidents involve ***Cars***.

# 16. How many accidents happened in London airport (Heathrow)?

In [171]:
UK[UK['District Area']=='London Airport (Heathrow)']['Index'].count()

np.int64(179)

## ***Insights:*** London Airport (Heathrow) has the fewest accidents with only **179 accidents** in the dataset.

# 17. How many accidents involved motorcycles over 500cc?

In [173]:
UK[UK['Vehicle_Type']=='Motorcycle over 500cc']['Index'].count()

np.int64(25657)

## ***Insights:*** There are **25,657 accidents** involving Motorcycles over 500cc, but most accidents still involve Cars.

# 18. What is the average latitude where accidents happened?

In [174]:
UK['Latitude'].mean()

np.float64(52.553865761110956)

## ***Insights:*** On average, accidents happened around latitude 52.55, which is roughly central in the UK.

# 19. What is the average longitude where accidents happened?

In [175]:
UK['Longitude'].mean()

np.float64(-1.431210368502073)

## ***Insights:*** On average, accidents happened around longitude -1.43, which is slightly west of the UK center.

# 20. How many accidents happened south of latitude 51?

In [176]:
UK[UK['Latitude'] < 51]['Index'].count()

np.int64(57288)

## ***Insights:*** There are **57,288 accidents** that happened south of latitude 51. This includes areas near the south coast of England.
